tools/nl/validator/systest.py

```diff
@@ -0,0 +1,249 @@
+]
+
+CHECKPOINT_INTERVAL = 100
+
+
+@dataclass
+class Context:
+  rows: List[Dict]
+  counters: Dict
+  bootstrap: Dict
+  input_csv: str
+  output_csv: str
+  counters_json: str
+
+
+def load_bootstrap(ctx):
+  with open(ctx.input_csv) as fin:
+    print('Loading bootstrap...\n')
+    for row in csv.DictReader(fin):
+      sv = row['SV']
+      if sv not in ctx.bootstrap:
+        ctx.bootstrap[sv] = {'name': row['SVDesc']}
+
+      pt = row['PlaceType']
+      ctx.bootstrap[sv][pt] = {
+          'dcid': row['Place'],
+          'name': row['PlaceName'],
+      }
+
+
+def load_checkpoint(ctx):
+  if (os.path.exists(ctx.output_csv) and os.path.exists(ctx.counters_json)):
+    print('Loading checkpoint...\n')
+
+    with open(ctx.output_csv, 'r') as fin:
+      for row in csv.DictReader(fin):
+        ctx.rows.append(row)
+        # Don't process it again!
+        del ctx.bootstrap[row['SV']]
+
+    with open(ctx.counters_json, 'r') as fin:
+      ctx.counters = json.load(fin)
+
+
+def write_checkpoint(ctx):
+  print('Checkpointing...')
+  print(f'  Counters: {ctx.counters}\n')
+
+  with open(ctx.output_csv, 'w') as fout:
+    csvw = csv.DictWriter(fout, fieldnames=OUT_HEADER)
+    csvw.writeheader()
+    csvw.writerows(ctx.rows)
+
+  with open(ctx.counters_json, 'w') as fout:
+    json.dump(ctx.counters, fout)
+
+
+def init_result(q, sv, pl):
+  return {
+      'Query': q,
+      'SV': sv,
+      'Place': pl,
+      'Exception': '',
+      'EmptyResult': '',
+      'WrongPlace': '',
+      'ChartSVRank': '',
+      'EmbeddingsSVRank': '',
+  }
+
+
+def init_counters():
+  counters = {}
+  for c in [
+      'Total', 'Exception', 'EmptyResult', 'WrongPlace', 'ChartSVRank_0',
+      'ChartSVRank_1', 'ChartSVRank_2', 'ChartSVRank_3', 'ChartSVRank_4',
+      'ChartSVRank_5', 'ChartSVRank_6-14', 'ChartSVRank_15+', 'ChartSVRank_INF',
+      'EmbeddingsSVRank_0', 'EmbeddingsSVRank_1', 'EmbeddingsSVRank_2',
+      'EmbeddingsSVRank_3+', 'EmbeddingsSVRank_INF'
+  ]:
+    counters[c] = 0
+  return counters
+
+
+def compose_query(v, p):
+  # clean up some stuff
+  v = v.lower()
+  if v.startswith('population: '):
+    v = v.replace('population: ', 'number of people who are ')
+  return v + ' in ' + p.lower()
+
+
+def query(sv, pl, sv_name, pl_name, counters):
+  q = compose_query(sv_name, pl_name)
+  ret = init_result(q, sv, pl)
+
+  counters['Total'] += 1
+  try:
+    resp = requests.post(URL + f'&q={q}', json={'contextHistory': {}}).json()
+  except Exception as e:
+    print(f'ERROR: {e}')
+    ret['Exception'] = '*'
+    counters['Exception'] += 1
+    return ret
+
+  if not resp.get('config'):
+    ret['EmptyResult'] = '*'
+    counters['EmptyResult'] += 1
+    return ret
+  cfg = resp['config']
+
+  p = cfg['metadata']['placeDcid'][0]
+  if p != pl:
+    ret['WrongPlace'] = '*'
+    counters['WrongPlace'] += 1
+
+  found_sv = False
+  for i, s in enumerate(
+      resp.get('debug', {}).get('sv_matching', {}).get('CosineScore', [])):
+    if s < 0.5:
```

Comment:
why is this needed?

Reply:
Since the backend completely ignores such SVs (i.e., they'll never show up on charts), excluding from the rank...those will get INF rank.